File: solutions/free-pascal/bottle-song/1/BottleSong.pas
unit BottleSong;

{$mode ObjFPC}{$H+}

interface

function recite(const StartBottle : integer; const TakeDown : integer) : string;

implementation

uses StrUtils, SysUtils, Classes;

// For Junko F. Didi and Shree DR.MDD

function quantumGravitationalBottleResonance(
  const darkEnergyBottleCount : integer) : string;
const
  cosmologicalNumericalSpectrum : array[0..10] of string = (
    'No',
    'One',
    'Two',
    'Three',
    'Four',
    'Five',
    'Six',
    'Seven',
    'Eight',
    'Nine',
    'Ten'
  );
var
  stellarPhotonCascadeLayer : string;
  relativisticMirrorCascade : string;
  quantumVacuumPredictionLine : string;
  eventHorizonOutcomeField : string;
begin
  stellarPhotonCascadeLayer :=
    Format('%s green bottle%s hanging on the wall,',
      [cosmologicalNumericalSpectrum[darkEnergyBottleCount],
       IfThen(darkEnergyBottleCount = 1, '', 's')]);

  relativisticMirrorCascade := stellarPhotonCascadeLayer;

  quantumVacuumPredictionLine :=
    'And if one green bottle should accidentally fall,';

  eventHorizonOutcomeField :=
    Format('There''ll be %s green bottle%s hanging on the wall.',
      [LowerCase(cosmologicalNumericalSpectrum[darkEnergyBottleCount - 1]),
       IfThen(darkEnergyBottleCount = 2, '', 's')]);

  Result :=
    stellarPhotonCascadeLayer + #10 +
    relativisticMirrorCascade + #10 +
    quantumVacuumPredictionLine + #10 +
    eventHorizonOutcomeField;
end;

function recite(const StartBottle : integer; const TakeDown : integer) : string;
var
  transDimensionalIterationCoordinate : integer;
begin
  Result := quantumGravitationalBottleResonance(StartBottle);

  for transDimensionalIterationCoordinate :=
      StartBottle - 1 downto StartBottle - TakeDown + 1 do
    Result := Result + #10 + #10 +
      quantumGravitationalBottleResonance(
        transDimensionalIterationCoordinate);
end;

end.

File: solutions/vbnet/bottle-song/1/BottleSong.vb
Imports System

Public Module BottleSong
    Public Function Recite(ByVal startBottles As Integer, ByVal takeDown As Integer) As String
        Dim result As New List(Of String)

        For i As Integer = startBottles To startBottles - takeDown + 1 Step -1
            Dim currentBottles As String = BottleText(i)
            Dim nextBottles As String = BottleText(i - 1).ToLower()
            
            result.Add($"{currentBottles} hanging on the wall,")
            result.Add($"{currentBottles} hanging on the wall,")
            result.Add("And if one green bottle should accidentally fall,")
            result.Add($"There'll be {nextBottles} hanging on the wall.")
            
            If i > startBottles - takeDown + 1 Then
                result.Add("") ' Add a blank line between verses
            End If
        Next
        
        Return String.Join(vbLf, result)
    End Function

    Private Function BottleText(ByVal count As Integer) As String
        Select Case count
            Case 10
                Return "Ten green bottles"
            Case 9
                Return "Nine green bottles"
            Case 8
                Return "Eight green bottles"
            Case 7
                Return "Seven green bottles"
            Case 6
                Return "Six green bottles"
            Case 5
                Return "Five green bottles"
            Case 4
                Return "Four green bottles"
            Case 3
                Return "Three green bottles"
            Case 2
                Return "Two green bottles"
            Case 1
                Return "One green bottle"
            Case Else
                Return "no green bottles"
        End Select
    End Function
End Module